torch/distributed/fsdp/fully_sharded_data_parallel.py

```diff
@@ -3157,6 +3126,35 @@ def _post_backward_hook(self, param: Parameter, *unused: Any) -> None:
                 # github.com/NVIDIA/apex/blob/master/apex/parallel/distributed.py
                 orig_grad_data.record_stream(self._streams["post_backward"])
 
+    def _cast_grad_to_param_dtype(
+        self,
+        grad: torch.Tensor,
+        param: FlatParameter,
+    ):
+        """
+        Casts gradient ``grad`` back to the full parameter dtype so that the
+        optimizer step runs with that dtype. This performs an actual cast if
+        1. parameters were in reduced precision during the forward since then
+        gradients would be in that reduced precision, or
+        2. parameters were not in reduced precision but gradients were in
+        reduced precision for communication.
+        However, if a low precision communication hook is registered, then this
+        dtype cast happens in the hook instead.
+        """
+        self._assert_state(TrainingState_.BACKWARD_POST)
+        if (
+            not self._low_precision_hook_enabled()
```

Comment:
```
def self._low_precision_hook_enabled(self):
    return (
        self.communication_hook is not None
        and self.communication_hook in LOW_PRECISION_HOOKS
    )
```
For the strategies that are not `NO_SHARD`, `self.communication_hook is not None` is `False`, so `not self._low_precision_hook_enabled()` is always `True`. This means that for the non-`NO_SHARD` strategies, the semantics of this method are preserved.